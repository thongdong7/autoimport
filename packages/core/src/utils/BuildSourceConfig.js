// @flow

/**
 * Scan source root to get exports
 */
import type { TImportInfo } from "../types";
import { withQueryTransform, isIdentifier } from "./Query";
import recast from "recast";
import type { TPackagesOption } from "../types";
import fs from "fs";
import path from "path";
import { prepareJscodeshift } from "./jscodeshift";

const j = prepareJscodeshift();

var types = recast.types.namedTypes;

const getExportNameVarIdentifier = (exportKind: "value" | "type") =>
  withQueryTransform({
    query: types.ExportNamedDeclaration,
    filter: {
      exportKind: exportKind,
    },
    flatNodeTransform: node => {
      const varDeclarations: Object[] =
        (node.declaration && node.declaration.declarations) || [];
      // console.log("declarations", declarations);

      const specifiers: Object[] = node.specifiers || [];
      const ret = [
        ...varDeclarations
          .filter(item => isIdentifier(item.id))
          .map(item => item.id.name),
        ...specifiers.map(item => item.exported.name),
      ];

      if (
        node.declaration &&
        (node.declaration.type === "FunctionDeclaration" ||
          node.declaration.type === "TypeAlias" ||
          node.declaration.type === "ClassDeclaration")
      ) {
        ret.push(node.declaration.id.name);
      }

      return ret;
    },
  });

const ExportNameVarIdentifier = getExportNameVarIdentifier("value");
const ExportNameTypeIdentifier = getExportNameVarIdentifier("type");

export function getExportInfo(file: string): TImportInfo {
  const source = fs.readFileSync(file).toString();

  return getExportInfoFromSource(file, source);
}

export function getExportInfoFromSource(
  file: string,
  source: string
): TImportInfo {
  const ast = j(source);
  const hasDefaultExport = ast.find(types.ExportDefaultDeclaration).size() > 0;

  const others = ExportNameVarIdentifier(ast);
  const otherTypes = ExportNameTypeIdentifier(ast);
  let ret: TImportInfo = {
    others,
    types: otherTypes,
  };

  if (hasDefaultExport) {
    let fileName = path.basename(file).replace(".js", "");
    if (fileName === "index") {
      fileName = path.basename(path.dirname(file));
    }

    ret.main = fileName;
  }

  return ret;
}

export function file2Path(file: string) {
  let ret = file.replace(".js", "");
  if (ret.endsWith("/index")) {
    ret = ret.substr(0, ret.length - 6);
  }

  return "./" + ret;
}

function* _getFiles(dir: string) {
  // There is no node_modules in rootPath (src folder)
  const dirName = path.basename(dir);
  // TODO Fix this hard-code (for ignore `build` dir)
  if (dirName === "node_modules" || dirName === "build") {
    return;
  }

  for (const file of fs.readdirSync(dir)) {
    const fullPath = path.join(dir, file);
    if (fs.lstatSync(fullPath).isDirectory()) {
      yield* _getFiles(fullPath);
    } else {
      yield fullPath;
    }
  }
}

export function scanSourceDir(rootPath: string): TPackagesOption {
  const packagesOptions: TPackagesOption = {};
  const jsFilePattern = /^\w+.js$/;

  if (fs.existsSync(rootPath)) {
    for (const file of _getFiles(rootPath)) {
      const fileName = path.basename(file);
      if (!jsFilePattern.test(fileName)) {
        continue;
      }

      const relativeFile = path.relative(rootPath, file);
      const relativePath = file2Path(relativeFile);

      const exportInfo = getExportInfo(file);

      if (
        exportInfo.main != null ||
        !exportInfo.others ||
        exportInfo.others.length > 0 ||
        !exportInfo.types ||
        exportInfo.types.length > 0
      ) {
        packagesOptions[relativePath] = exportInfo;
      }
      // optionSource[_file2Path(relativeFile)] = importInfo;
    }
  }

  return packagesOptions;
}
