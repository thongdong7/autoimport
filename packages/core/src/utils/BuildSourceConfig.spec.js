// @flow

import { getExportInfoFromSource, scanSourceDir } from "./BuildSourceConfig";
import type { TImportInfo } from "../types";
import path from "path";
import fs from "fs";

describe("getExportInfoFromSource", () => {
  const defaultFile = "components/A.js";

  const d = path.join(__dirname, "../../__testfixtures__/DetectExport");
  /* eslint-disable no-unused-vars */
  function codeFile(file) {
    /* eslint-enable no-unused-vars */
    const source = fs.readFileSync(path.join(d, file)).toString("utf-8");
    return code(source, file);
  }

  function code(source: string, file?: string) {
    const filePath = file != null ? file : defaultFile;

    const importInfo: TImportInfo = getExportInfoFromSource(filePath, source);

    const checker = {
      hasOthers(...members: string[]) {
        const others = importInfo.others || [];

        for (const member of members) {
          expect(others).toContain(member);
        }

        return checker;
      },
      hasOtherTypes(...members: string[]) {
        const others = importInfo.types || [];

        for (const member of members) {
          expect(others).toContain(member);
        }

        return checker;
      },
      hasMain(member: string) {
        expect(importInfo.main).toBe(member);

        return checker;
      },
      notHasMain() {
        expect(importInfo.main).toBeUndefined();

        return checker;
      },
    };

    return checker;
  }

  // fit("debug", () => {
  //   codeFile("debug.js")
  //     .hasOthers("A")
  //     .notHasMain();
  // });

  it("export function", () => {
    code(`
      export function A();
    `)
      .hasOthers("A")
      .notHasMain();
  });

  it("export idenfiers", () => {
    code(`
      export const A = 1;
      export let B = 2, C=3;
    `)
      .hasOthers("A", "B", "C")
      .notHasMain();
  });

  it("export identifier as another name", () => {
    code(`
      const A = 1;
      const B = 2;

      export {A, B as B1};
    `)
      .hasOthers("A", "B1")
      .notHasMain();
  });

  it("export default in non-index file will be file name", () => {
    code(
      `
      export default 1;
    `,
      "/ABC.js"
    ).hasMain("ABC");
  });

  it("export default in index.js will be detected name base on folder name", () => {
    code(
      `
      export default 1;
    `,
      "/ABC/index.js"
    ).hasMain("ABC");
  });

  it("export type and value", () => {
    code(
      `
    export type TUser = {
      name: string,
    }

    export const A = 1;
    export default 2;
    `,
      "/ABC.js"
    )
      .hasOtherTypes("TUser")
      .hasMain("ABC")
      .hasOthers("A");
  });
});

describe("scanSourceDir", () => {
  it("could scan dir", () => {
    const d = path.join(__dirname, "../../__testfixtures__/sample1");

    const packagesFromSource = scanSourceDir(d);
    // console.log(typeOptionsSource);
    expect(packagesFromSource["./A"]).toBeDefined();
    expect(packagesFromSource["./B"]).toBeDefined();

    expect(packagesFromSource).toMatchSnapshot();

    expect(packagesFromSource["./B/types"]).toBeDefined();
    expect(packagesFromSource).toMatchSnapshot();
  });
});
