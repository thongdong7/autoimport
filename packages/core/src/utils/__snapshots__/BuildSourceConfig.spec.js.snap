// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scanSourceDir could scan dir 1`] = `
Object {
  "./A": Object {
    "main": "A",
    "others": Array [
      "x",
    ],
    "types": Array [],
  },
  "./B": Object {
    "main": "B",
    "others": Array [],
    "types": Array [],
  },
  "./B/CNoDefault": Object {
    "others": Array [
      "C",
    ],
    "types": Array [],
  },
  "./B/types": Object {
    "others": Array [],
    "types": Array [
      "TUser",
    ],
  },
}
`;

exports[`scanSourceDir could scan dir 2`] = `
Object {
  "./A": Object {
    "main": "A",
    "others": Array [
      "x",
    ],
    "types": Array [],
  },
  "./B": Object {
    "main": "B",
    "others": Array [],
    "types": Array [],
  },
  "./B/CNoDefault": Object {
    "others": Array [
      "C",
    ],
    "types": Array [],
  },
  "./B/types": Object {
    "others": Array [],
    "types": Array [
      "TUser",
    ],
  },
}
`;
